src/vos/vos_dtx.c

```diff
@@ -717,6 +730,469 @@ vos_dtx_abort_one(struct vos_container *cont, struct daos_tx_id *dti,
+	rec->tr_type = type;
+	rec->tr_flags = flags;
+	rec->tr_record = record;
+	rec->tr_next = UMMID_NULL;
+
+	if (UMMID_IS_NULL(dth->dth_ent)) {
+		struct vos_container	*cont;
+		umem_id_t		 dtx_mmid;
+		daos_iov_t		 kiov;
+		daos_iov_t		 riov;
+		struct dtx_rec_bundle	 rbund;
+
+		D_ASSERT(flags == 0);
+
+		cont = vos_hdl2cont(dth->dth_coh);
+		D_ASSERT(cont != NULL);
+
+		dtx_mmid = umem_zalloc(umm, sizeof(struct vos_dtx_entry_df));
+		if (UMMID_IS_NULL(dtx_mmid))
+			return -DER_NOMEM;
+
+		dtx = umem_id2ptr(umm, dtx_mmid);
+		dtx->te_xid = dth->dth_xid;
+		dtx->te_oid = dth->dth_oid;
+		if (dth->dth_dkey != NULL) {
+			dtx->te_dkey_hash[0] = d_hash_murmur64(
+						dth->dth_dkey->iov_buf,
+						dth->dth_dkey->iov_len,
+						VOS_BTR_MUR_SEED);
+			dtx->te_dkey_hash[1] = d_hash_string_u32(
+						dth->dth_dkey->iov_buf,
+						dth->dth_dkey->iov_len);
+		} else {
+			dtx->te_dkey_hash[0] = 0;
+			dtx->te_dkey_hash[1] = 0;
+		}
+		dtx->te_epoch = dth->dth_epoch;
+		dtx->te_ver = dth->dth_ver;
+		dtx->te_state = DTX_ST_INIT;
+		dtx->te_flags = 0;
+		dtx->te_intent = dth->dth_intent;
+		dtx->te_sec = time(NULL);
+		dtx->te_records = rec_mmid;
+		dtx->te_next = UMMID_NULL;
+		dtx->te_prev = UMMID_NULL;
+
+		rbund.trb_ummid = dtx_mmid;
+		daos_iov_set(&riov, &rbund, sizeof(rbund));
+		daos_iov_set(&kiov, &dth->dth_xid, sizeof(dth->dth_xid));
+		rc = dbtree_upsert(cont->vc_dtx_active_hdl,
+				BTR_PROBE_EQ, DAOS_INTENT_UPDATE, &kiov, &riov);
+		if (rc != 0)
+			return rc;
+
+		dth->dth_ent = dtx_mmid;
+	} else {
+		/* The @dtx must be new created via former
+		 * vos_dtx_register_record(), no need umem_tx_add_ptr().
+		 */
+		dtx = umem_id2ptr(umm, dth->dth_ent);
+		dtx->te_sec = time(NULL);
+		rec->tr_next = dtx->te_records;
+		dtx->te_records = rec_mmid;
+
+		if (flags != 0) {
+			struct vos_dtx_record_df	*tgt;
+
+			/*
+			 * XXX: Currently, we only support DTX_RF_EXCHANGE_SRC
+			 *	when register the target for punch {a,d}key.
+			 *	It is implemented as exchange related {a,d}key.
+			 *	The exchange target has registered its record
+			 *	to the DTX just before the exchange source.
+			 */
+			D_ASSERT(flags == DTX_RF_EXCHANGE_SRC);
+			D_ASSERT(type == DTX_RT_OBJ || type == DTX_RT_KEY);
+
+			/* The @tgt must be new created via former
+			 * vos_dtx_register_record(), no need umem_tx_add_ptr().
+			 */
+			tgt = umem_id2ptr(umm, rec->tr_next);
+			D_ASSERT(tgt->tr_flags == 0);
+
+			tgt->tr_flags = DTX_RF_EXCHANGE_TGT;
+			dtx->te_flags |= DTX_EF_EXCHANGE_PENDING;
+
+			if (type == DTX_RT_OBJ) {
+				struct vos_obj_df	*obj;
+
+				obj = umem_id2ptr(umm, record);
+				obj->vo_oi_attr |= VOS_OI_REMOVED;
+			} else {
+				struct vos_krec_df	*key;
+
+				key = umem_id2ptr(umm, record);
+				key->kr_bmap |= KREC_BF_REMOVED;
+			}
+
+			D_DEBUG(DB_TRACE, "Register exchange source "
+				"for %s DTX "DF_DTI"\n",
+				type == DTX_RT_OBJ ? "OBJ" : "KEY",
+				DP_DTI(&dtx->te_xid));
+		}
+	}
+
+	*entry = dth->dth_ent;
+	if (shares != NULL)
```

Comment:
I'm a little confused, why it's set to 1?

Reply:
"1" means that there is one (uncommitted) DTX referencing such target. If there are more DTXs share such target later, then the @shares will be increased.